### PR TITLE
Fix handling of Hs in SMARTS queries

### LINKED ISSUE #1338: SMARTS proton query parsed incorrectly
This looks like a bug (from the mailing list):

>>> q = Chem.MolFromSmarts("[H+]")
>>> for atom in q.GetAtoms():
...   print atom.DescribeQuery()
... 
AtomAnd
  AtomHCount 1 = val
  AtomFormalCharge 1 = val

### PR DESCRIPTION
**Pull request by greglandrum**
The change is relatively small, but I find the "documentation" about what should happen here somewhat ambiguous. Please take a look at the tests I added to see if you agree that the right thing is now happening.

Fixes #1338

### CODE REVIEW

**Comment by johnmay** [comment]
Heard this on the grapevine (via Noel). With a quick skim, it looks like the semantics are almost there but not quite. The rule of thumb is for hydrogen semantics in SMARTS if it's a valid SMILES then interpret it as an element otherwise it's total hydrogen count.

```
[H] == [#1]
[H+] == [#1+]
[2H] == [2#1]
[2H+] == [2#1+]
[N;H+] != [N;#1+]
```

**Comment by bp-kelley** [comment]
@johnmay I ping Noel, mainly because I expected he wouldn't let this one go either ;)

**Comment by greglandrum** [comment]
Thanks @johnmay 
Do you have a pointer to where that rule of thumb comes from?

**Comment by johnmay** [comment]
> These differences in interpretation would make SMIRKS unintelligible. Hence, a single change to SMARTS interpretation, for expressions of the form: [<weight>]H<charge><map>]. In SMARTS, these expressions now are interpreted as a hydrogen atom, rather than as any atom with one hydrogen attached. All other SMARTS hydrogen expressions retain their pre-4.51 meanings.

http://www.daylight.com/dayhtml/doc/theory/theory.smirks.html

**Comment by johnmay** [comment]
...reminds me I need to fix this in CDK also, on the long ToDo list

**Comment by greglandrum** [comment]
I guess that's at least simpler to parse, but it sure feels wrong

**Comment by greglandrum** [comment]
@bp-kelley : I need to fix the code to make this work.
@johnmay : thanks for the explanation and pointer

**Comment by greglandrum** [comment]
OpenSMARTS is going to be soooo much fun! :-S

**Comment by johnmay** [comment]
Yeah the ambiguity is hinted at but not resolved: https://github.com/timvdm/OpenSMARTS/blob/master/source/reading.rst#total-hydrogen-count

**Comment by greglandrum** [comment]
@bp-kelley : I believe the new form of the parser does what @johnmay described. There are some tests in there for it too.
This should be once again ready for review.

**Comment by bp-kelley** [comment]
Thanks greg, since this might affect ReactionSmarts I might throw in an additional couple of tests there as well.  It will take the weekend.

File: Code/GraphMol/SmilesParse/smatest.cpp [hunk 1]
```diff
@@ -1865,6 +1864,164 @@ void testGithub893() {
...
+    delete p;
+  }
+  {
+    RWMol *p;
+    std::string sma = "[2H:3]";
+    p = SmartsToMol(sma);
+    TEST_ASSERT(p);
+
+    TEST_ASSERT(p->getAtomWithIdx(0)->getAtomicNum() == 1);
+    TEST_ASSERT(p->getAtomWithIdx(0)->getIsotope() == 2);
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/smatest.cpp]: Might as well check the atom map...

**Comment by bp-kelley** [comment]
So, these are absolutely meaningless in practice, but it does show some holes in how MergeQueryHs works.  I'm not sure that this is related, but could be "fixed" using the above semantics.

Insane example:

```
>> m = Chem.MolFromSmiles("C[H+]")
>>> Chem.MolToSmarts(Chem.MergeQueryHs(m))
'[#6H]'
```

Slightly more reasonable:

```
>>> m = Chem.MolFromSmiles("C[2H]")
>>> m2 = Chem.MolFromSmiles("C[H]")
>>> m2.HasSubstructMatch(m)
False
>>> m.HasSubstructMatch( Chem.MergeQueryHs(m) )
True
>>> m2.HasSubstructMatch( Chem.MergeQueryHs(m) )
True
>>> 
```

**Comment by greglandrum** [comment]
good point on `MergeQueryHs()`: that should be modified to not merge in something with an isotope or a charge.
I'm going to do a separate issue for that though; it's not part of the way SMARTS parsing is handled.

File: Code/GraphMol/SmilesParse/smatest.cpp (hunk 1 above)
Comment (greglandrum) [inline Code/GraphMol/SmilesParse/smatest.cpp]: good point; new commit coming.

**Pull request merged**